fvm/src/call_manager/default.rs

```diff
@@ -51,6 +52,8 @@ pub struct InnerDefaultCallManager<M: Machine> {
     invocation_count: u64,
     /// Limits on memory throughout the execution.
     limits: M::Limiter,
+    /// Accumulator for events emitted in this call stack.
+    events: Vec<StampedEvent>,
```

Comment:
Yeah, this is unfortunately a bit of a can of worms.

1. Filecoin does not revert side effects on non-zero exit codes (i.e. you can change your state root and abort with a non-zero exit code, and the protocol doesn’t revert your state change).
2. Events are trickier though, because they are fire-and-forget. Once an event is emitted, you can't take a compensating action to un-emit it.
    - By contrast, you can always revert a state root update. For example, you track your previous state root, update your state root, detect a failure scenario, and revert back to your original state root, before returning a non-zero exit code.
3. Ethereum _does_ revert side effects (including storage updates) on REVERT. We currently enforce the rollback of storage mutations by using a transaction block in the EVM runtime (i.e. entirely actor-side), but this begs to introduce protocol support for rollback on failure, which would create a new precedent.

I can think of two ways to handle this:

1. Layer events in the call manager and the last layer on a non-zero exit. This would prevent us from intentionally emitting events on failure (e.g. I could reasonably conceive a "window post failed" event).
2. Enable actors to throw away events emitted through their execution. The EVM runtime would use this feature, and we could make the transactional Runtime discard events by default on failure. Future native actors (or built-in actors) could choose retain events on failure.

Reply:
No, Filecoin totally reverts everything on non-zero exit codes.

Reply:
^What Steb said -- we do so in both [the legacy VM](https://github.com/filecoin-project/lotus/blob/8df5b81f24d66c82d409d57a1ccc437e5f0d2465/chain/vm/vm.go#L562-L567) and [the FVM](https://github.com/filecoin-project/ref-fvm/blob/0801fd8f523692c0d17cf8c41fae2143bcff332a/fvm/src/call_manager/default.rs#L167).

I think that makes most (all?) of the problems you're describing go away.

Reply:
I was in doubt and it sounded strange that it didn't, but I wasn't able to find where it was happening. And that happens through the layering in the state tree: https://github.com/filecoin-project/ref-fvm/blob/0801fd8f523692c0d17cf8c41fae2143bcff332a/fvm/src/call_manager/default.rs#L160